Added command_line auth provider that validates credentials by calling a command

bob1de opened the pull request:
## Description:

A new authentication provider that checks username/password by calling an external program, passing the values as environment variables. If the program exits with exit code 0, authentication succeeds.

Additionally, the program can print out lines of the form

    KEY=VALUE

to give HA some information about the authenticating user. Currently, only ``name`` is supported, but groups are thinkable as well when there will be an official API for interfacing with HA's group system in the future.

As an example, I wrote a script to do authentication against LDAP, that I'll link here soon.

**Related issue (if applicable):** Closes #19975

**Pull request in home-assistant-polymer:** home-assistant/home-assistant-polymer#2561

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8321

## Example entry for `configuration.yaml` (if applicable):
```yaml
homeassistant:
  auth_providers:
  - type: command_line
    command: /config/auth.sh
    # Optionally, a list of arguments.
    #args: []
    # Optionally, accept meta variables.
    #meta: true
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code does not interact with devices:
  - [X] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Commit:
Added external auth provider that calls a configurable program

Closes #19975

Commit:
Raise proper InvalidAuth exception on OSError during program execution

balloob commented on homeassistant/auth/providers/external.py:
```diff
@@ -0,0 +1,149 @@
+    """Raised when authentication with given credentials fails."""
+
+
+@AUTH_PROVIDERS.register("external")
```
Name is ambiguous, all auth providers besides "homeassistant" are external…

I would suggest "command_line"

efficiosoft replied:
Agreed, will change that tomorrow.

Commit:
Changed name of external auth provider to command_line

awarecan commented on homeassistant/auth/providers/command_line.py:
```diff
@@ -0,0 +1,149 @@
+                line = _line.decode().lstrip()
+                if line.startswith("#"):
+                    continue
+                key, value = line.split("=", 1)
```
This way to grab metadata is too specific for your usage, I would suggest generalize it by using template(possible?) or regex match

efficiosoft replied:
What kind of metadata do you think this is not suited for? The only thing that's allowed to be set by HA right now is the name, and there will probably be groups in the future, but what else?

balloob replied:
I think that templates and regex are a bit too error prone, and this is sensitive matter once it includes groups. So I suggest that we just define a format, and I think in that case the KEY=VALUE format is ok? Especially if it's a bash script, it might not be easy to define a more complicated output.

efficiosoft replied:
Yes, exactly. And instead of having to maintain regexes in the HA configuration and trust they're applied as I expect, I feel much more comfortable with adding the formatting directly to the adapter that integrates my auth backend. I think no one will use something like ``ldapsearch`` directly without a script around.

Maybe we could make it configurable whether to read meta variables from stdout or not, in case somebody really wants to use something he can't control the output of. What about a ``meta: false`` option which you have to turn on explicitly? Could easily be described in the docs.

@balloob What do you mean with MVP?

efficiosoft replied:
Agreed on ``meta`` parameter and current implementation as discussed below. Marking this resolved.

awarecan reviewed (changes requested):
Please add unit test.

You can include a script to mock external program in your unit test to return valid or invalid result

efficiosoft commented:
I'll create an unit test when the API is final.

balloob commented:
This looks good. Will need some tests written.

balloob commented:
I suggest that for MVP we remove the metadata and just look at the return code. Keep it simple.

efficiosoft commented:
Ah I see, MVP... Removing the metadata sounds not desirable to me. I don't want to see "You're logged in as ." in the UI because HA doesn't know my name. Could become more problematic when HA maybe logs user actions some day.

Having to turn it on explicitly sounds more appropriate to me.

Commit:
Renamed program config option to command in command_line auth provider

Commit:
Made meta variable parsing in command_line auth provider optional

Commit:
Added tests for command_line auth provider

Commit:
Fixed indentation

efficiosoft commented:
And here's a working example for authentication against LDAP:

https://github.com/efficiosoft/ldap-auth-sh

Commit:
Suppressed wrong pylint warning

Commit:
Fixed linting

frenck commented:
Could not find a matching documentation PR on our documentation repository, adding `docs-missing` label.

awarecan commented on homeassistant/auth/providers/command_line.py:
```diff
@@ -0,0 +1,160 @@
+            raise InvalidAuthError
+
+        if process.returncode != 0:
+            raise InvalidAuthError
```
Better add an error log with return code

awarecan commented on homeassistant/auth/providers/command_line.py:
```diff
@@ -0,0 +1,160 @@
+        }
+        try:
+            # pylint: disable=no-member
+            process = await asyncio.subprocess.create_subprocess_exec(
```
`asyncio.subprocess` already imported

efficiosoft replied:
pylint mixes ``asyncio.subprocess`` with ``subprocess`` and complains about missing ``create_subprocess_exec`` therein, don't know why. I imported the right module.

awarecan commented on homeassistant/auth/providers/command_line.py:
```diff
@@ -0,0 +1,160 @@
+"""Auth provider that validates credentials via an external command."""
+
+import typing as T
```
I think this is not following our standard, could change to something likes
```
from typing import Any, Dict, List, Optional, cast
```

efficiosoft replied:
Maybe not like it has been done for the other auth providers, but I find it awkward to mix the typing classes into global namespace. This really pollutes it, as there can be quite a lot of them laying around. It's much cleaner to hide all of this functionally unrelated stuff behind a single ``T``, imho.

efficiosoft replied:
So what do you say, is changing it a strict requirement for the PR to get accepted, or do you see the advantages of not throwing every single type class into module namespace like I do?

awarecan replied:
I cannot recall other files are using your style. Coding style could be very personal, however, we have to follow the norms in order to maintenance a readable code base.

awarecan reviewed (changes requested):
There are two stuffs are missing besides the documentation pointed by Frenck
* unit test for login flow
* a front-end PR to include string resources, you can use leagcy_api_password auth provider as example: https://github.com/home-assistant/home-assistant-polymer/blob/f98fff9ffdd101b029f44e8ffe5a59545d388e84/src/translations/en.json#L939

efficiosoft commented on homeassistant/auth/providers/command_line.py:
```diff
@@ -0,0 +1,160 @@
+    # which keys to accept from a program's stdout
+    ALLOWED_META_KEYS = ("name",)
+
+    def __init__(self, *args: T.Any, **kwargs: T.Any) -> None:
```
Would be a member of the type then, not of the instance... in terms of object isolation, that sound's not desirable, even it would make no difference as long as only ONE single instance of the auth provider exists at the same time - but I'm against it for said reasons.

awarecan replied:
Aha, you are damn right.

efficiosoft commented:
@frenck and @awarecan 

I won't create any docs until we mutually agreed on the API, won't do it twice when it needs to be changed. Are you happy with ``meta`` being optional and ``False`` by default?

I'll look into testing the login flow as well... Just took the existing tests from ``insecure_example``, which didn't include the login flow.

Strings for the frontend... alright.

awarecan commented:
I am fine with meta being optional and False by default

awarecan commented:
Here is the example of login flow test, https://github.com/home-assistant/home-assistant/blob/201fd4afeec56f031dab027f0b84debea5bced77/tests/auth/providers/test_homeassistant.py#L67

efficiosoft commented:
Yes, the ``homeassistant`` provider has one. Will cope with this tomorrow.

frenck commented:
@efficiosoft About the docs, that is fine, nevertheless, as long as the docs aren't in the label should be up here to remind code reviewers of missing documentation...

Commit:
Added test for command line auth provider login flow

Commit:
Log error when user fails authentication

Commit:
Use %r formatter instead of explicit repr()

Commit:
Mix all used names of typing module into module namespace

I consider this nasty and bad coding style, but was requested by
@awarecan for consistency with the remaining codebase.

efficiosoft commented:
I've just written the docs and am wondering what the policy regarding external links is. I would like to link the script for LDAP authentication I maintain as an example application for the new auth provider.

efficiosoft commented on homeassistant/auth/providers/command_line.py:
```diff
@@ -0,0 +1,163 @@
+                stdout=asyncio.subprocess.PIPE
+                if self.config[CONF_META] else None,
+            )
+            stdout = (await process.communicate())[0]
```
You want each line of stderr to be logged via the logging system in case authentication fails? At which level?

What should happen with stderr when authentication succeeds? I ask because it's sometimes used for status messages that are no errors.

efficiosoft replied:
Right now, stderr always goes to stdout of the HA process.

awarecan replied:
I see, stderr was ignored in `create_subprocess_exec`, so this line can be changed a little bit 
```python
stdout, _ = await process.communicate()
```

klaasnicolaas commented:
https://github.com/home-assistant/home-assistant.io/pull/8321

Commit:
Small code style change

efficiosoft commented:
I think this PR is ready for final review and merge now.

balloob commented on homeassistant/auth/providers/command_line.py:
```diff
@@ -0,0 +1,163 @@
+    async def async_validate_login(self, username: str, password: str) -> None:
+        """Validate a username and password."""
+        env = {
+            "username": username,
```
Should we strip the username?

balloob commented on homeassistant/auth/providers/command_line.py:
```diff
@@ -0,0 +1,163 @@
+            self, flow_result: Dict[str, str]
+    ) -> Credentials:
+        """Get credentials based on the flow result."""
+        username = flow_result["username"]
```
Should we strip the username?

(usernames are not stripped before being passed on because we didn't do it initially, and so doing so now will be a breaking change. We have put a guard in place to ensure it won't happen and we will start stripping usernames in 2020)

balloob reviewed (approved):
It looks good 🎉  Just one question about the username stripping. Probably easiest to implement the username stripping in the config flow btw.

efficiosoft commented:
@balloob Hmm, I'm not sure I understand correctly. Usernames aren't stripped by this auth provider, as with any other existing provider I think.

efficiosoft commented:
The only things that are stripped are the meta variables(currently only the user's real name).

awarecan commented:
Balloob is referencing a recent merged PR, #20150

balloob commented:
If a user enters `"hello "`, and the next time it enters `"hello"`, the command line will probably see it as the same (or not?), the credentials are matched on user input, not on what script understood of it being.

efficiosoft commented:
No, it won't. Environment variables preserve all characters. However, it depends on the script the user configures to run to not strip them itself. I can add a note about that to the docs.

efficiosoft commented:
Did it.

efficiosoft commented:
On the other hand, I see nothing that stands against stripping usernames from the beginning on. So do you think it should simply be done before it get's released to prevent having a breaking change later?

efficiosoft commented:
@balloob Considering the modernized homeassistant auth provider from #20150, I'd probably start stripping usernames with the command_line auth provider from the beginning. Will change it in code & docs.

Commit:
Strip usernames with command_line auth provider

balloob commented:
Niiiiiice 🎉
